---
title: "Alpha-diversity-LME"
author: "Katherine Hinton"
output: html_document
date: "2025-08-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(knitr)
library(microbiome) # For alpha-diversity measures
library(phytools)
library(picante) # For Faith's PD
library(ape) # For phylotree
library(dplyr)
library(phyloseq) # Midpoint root 
library(lme4) # Linear model
library(lmerTest)     # v3.1.3 for lmer with p-values
library(car)          # v3.1.2 for Anova (Type II)
library(performance)  # v0.12.0 for R2 and residual checks
library(DHARMa) 
library(RColorBrewer) # ggplot
library(data.table)
library(emmeans) # Post hoc test
library(tidyverse) # Pivot longer
library(microbiomeutilities) # theme_biome_utils
library(Biostrings)
library(vegan)
library(rstatix)

options(bitmapType='cairo') # Visualise plots on interactive R
```

Load data using phyloseq
```{r, include=FALSE}
#read metadata
dataset <- read.csv("/rds/homes/k/kgh742/psf_wgs_project/02.MicrobiomeAnalysis/Metadata_microbiome.csv")
rownames(dataset) <- dataset$Sample_name        ##Rownames may not be recognised (used sample_names() to check)
dataset[] <- lapply(dataset, as.factor)

#read taxa, if starting fro here
taxa <- (read.csv("/rds/homes/k/kgh742/psf_wgs_project/02.MicrobiomeAnalysis/16S_Raw/tax_table_16S.csv"))
taxa <- taxa %>% 
  tibble::column_to_rownames("X")
taxa <-as.matrix(taxa)

#load ASV table, if starting from here
library(data.table)
seqtab.nochim <- fread("/rds/homes/k/kgh742/psf_wgs_project/02.MicrobiomeAnalysis/16S_Raw/ASV_table_16s.csv")
seqtab.nochim <- seqtab.nochim %>% 
  tibble::column_to_rownames("V1")
seqtab.nochim <- otu_table(seqtab.nochim , taxa_are_rows=FALSE)

##Create phyloseq object. Sample data needs to provide information about the samples
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), sample_data(dataset), tax_table(taxa))

##Assign the taxa names instead of DNA sequences to the OTU table and DNA sequneces as reference sequences
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names <- paste0("ASV", seq(ntaxa(ps)))
ps

#subsetting to remove mitochondria and chloroplast
ps.clean = subset_taxa(ps, Order != "Chloroplast" & Family != "Mitochondria")

tree <- read.tree("/rds/homes/k/kgh742/psf_wgs_project/02.MicrobiomeAnalysis/16S_fasttree_030225.nwk")
print(tree)

# Tree must be rooted for Fait's PD index. Root the tree using midpoint rooting. I used a random outgroup from my dataset.
tree_rooted <- midpoint.root(tree) # Root the tree

is.rooted(tree_rooted) #check if it worked

ps.clean <- merge_phyloseq(ps.clean, phy_tree(tree_rooted)) #merge to phyloseq object 

pseq <- ps.clean

ps_filtered0 <- prune_samples(sample_sums(pseq) >= 1000, pseq)

sample_data(ps_filtered0)$Site_Tree <- factor(paste(sample_data(ps_filtered0)$Site, sample_data(ps_filtered0)$Tree, sep = "_"))

```


```{r, include=FALSE}
# filter to remove any ASV with <10 reads
ps1 <- prune_taxa(taxa_sums(ps_filtered0) >= 10, ps_filtered0)

#make dataframe with the selected indices, plus the phylogenetic diversity index
adiv <- data.frame(  "Observed" = microbiome::alpha(ps1, index = "observed"),
                     "Shannon"= microbiome::alpha(ps1, index = "diversity_shannon"),
                    "dominance_dbp"= microbiome::alpha(ps1, index = "dominance_dbp"),
                    "InvserseSimpson" = microbiome::alpha(ps1, index  = "diversity_inverse_simpson"),
                    "EvennessPielou"= microbiome::alpha(ps1, index = "evenness_pielou"),
                    "Phylogenetic" = picante::pd(samp = as.data.frame(phyloseq::otu_table(ps1)),
                     tree = phyloseq::phy_tree(ps1), include.root=TRUE))

setDT(adiv, keep.rownames = "Sample_name")

ps1.meta<- merge(data.frame(sample_data(ps1)), adiv, by = "Sample_name")

# Save read depth (based on full ASV table) for use as covariate in the model
ps1.meta$ReadDepth <- sample_sums(ps_filtered0)
ps1.meta$ReadDepth_z <- scale(ps1.meta$ReadDepth)  # z-score standardization

# Specify tissue health reference
ps1.meta$Tissue_health <- factor(ps1.meta$Tissue_health, levels = c("Symptomatic", "Healthy", "Non-symptomatic"))
# Check Tree factor levels
levels(ps1.meta$Tissue_health)
```


Run linear model, anova, emmeans post-hoc and tests for normality on each alpha-metric (using analyse_alpha_diversity function)

```{r}
res_shannon <- analyse_alpha_diversity(ps1.meta, "diversity_shannon")
res_FPD <- analyse_alpha_diversity(ps1.meta, "Phylogenetic.PD", pairwise = TRUE)
res_observed <- analyse_alpha_diversity(ps1.meta, "observed", pairwise = TRUE)
res_evenness <- analyse_alpha_diversity(ps1.meta, "evenness_pielou")
res_DDBP <- analyse_alpha_diversity(ps1.meta, "dominance_dbp")
```

Save ANOVA and post-hoc emmeans results into dataframe for supplementary figures (use results_to_df function)
```{r}
# Combine ANOVA results
anova_all <- bind_rows(
  results_to_df(res_shannon, "Shannon diversity"),
  results_to_df(res_FPD, "Faith's PD"),
  results_to_df(res_DDBP, "Dominance DBP"),
  results_to_df(res_evenness, "Evenness Pielou"),
  results_to_df(res_observed, "Observed richness")
)

# Combine emmeans results
emmeans_all <- bind_rows(
  results_to_df(res_shannon, "Shannon diversity", emmeans = TRUE),
  results_to_df(res_FPD, "Faith's PD", emmeans = TRUE),
  results_to_df(res_DDBP, "Dominance DBP", emmeans = TRUE),
  results_to_df(res_evenness, "Evenness Pielou", emmeans = TRUE),
  results_to_df(res_observed, "Observed richness", emmeans = TRUE)
)
```

PLOT
```{R}

ps1_long <- ps1.meta %>%
  pivot_longer(
    cols = c("observed", "diversity_shannon", "Phylogenetic.PD"),  # Choose alpha metrics of interest
    names_to  = "Index",
    values_to = "Value") %>% 
  mutate(Tissue_health = factor(Tissue_health, levels = c("Healthy", "Non-symptomatic", "Symptomatic")))

ggplot(ps1_long, aes(
  x = factor(Site,
             levels = c("Lathkill_dale", "Wytham"),
             labels = c("Lathkill Dale", "Wytham Woods")),
  y = Value,
  fill = Tissue_health
)) +
  geom_boxplot(
    position = position_dodge(width = 0.8),
    outlier.size = 0.6,
    outlier.colour = "red"
  ) +
  geom_jitter(
    position = position_jitterdodge(jitter.width = 0.1, dodge.width = 0.8),
    size = 0.6,
    alpha = 0.6
  ) +
  facet_wrap(
    ~ Index,
    scales = "free_y",
    labeller = labeller(Index = c(
      diversity_shannon = "Shannon Diversity",
      observed = "Observed Richness",
      Phylogenetic.PD = "Faith's Phylogenetic Diversity"
    ))
  ) +
  scale_fill_manual(values = c("#2E4F79", "#93D6A0", "pink")) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  labs(x = "Site", y = "Diversity Index Value") +
  theme_biome_utils() +
  theme(
    axis.text.x = element_text(size = 10, angle = 45, hjust = 1),
    axis.text.y = element_text(size = 10),
    legend.position = "top",
    text = element_text(size = 10)
  )

```
